test/e2e/network/ingress_scale.go

```diff
@@ -0,0 +1,329 @@
+	ingressesCleanupTimeout        = 60 * time.Minute
+)
+
+var (
+	scaleTestLabels = map[string]string{
+		"app": scaleTestBackendName,
+	}
+)
+
+// Limit the number of parallel ingress creation.
+var threadLimit = make(chan struct{}, maxConccurentIngressCreations)
+
+func acquireThread() {
+	threadLimit <- struct{}{}
+}
+
+func releaseThread() {
+	<-threadLimit
+}
+
+var _ = SIGDescribe("Loadbalancing: L7 Scalability", func() {
+	defer GinkgoRecover()
+	var (
+		ns string
+	)
+	f := framework.NewDefaultFramework("ingress-scale")
+
+	BeforeEach(func() {
+		ns = f.Namespace.Name
+	})
+
+	Describe("GCE [Slow] [Serial] [Feature:IngressScale]", func() {
+		var (
+			jig           *framework.IngressTestJig
+			gceController *framework.GCEIngressController
+			scaleTestSvcs []*v1.Service
+			scaleTestIngs []*extensions.Ingress
+		)
+
+		BeforeEach(func() {
+			framework.SkipUnlessProviderIs("gce", "gke")
+
+			By("Initializing ingress test suite and gce controller")
+			jig = framework.NewIngressTestJig(f.ClientSet)
+			gceController = &framework.GCEIngressController{
+				Ns:     ns,
+				Client: jig.Client,
+				Cloud:  framework.TestContext.CloudConfig,
+			}
+			gceController.Init()
+
+			scaleTestSvcs = []*v1.Service{}
+			scaleTestIngs = []*extensions.Ingress{}
+		})
+
+		AfterEach(func() {
+			By("Cleaning up services...")
+			for _, svc := range scaleTestSvcs {
+				if svc != nil {
+					jig.TryDeleteGivenService(svc)
+				}
+			}
+			By("Cleaning up ingresses...")
+			for _, ing := range scaleTestIngs {
+				if ing != nil {
+					jig.TryDeleteGivenIngress(ing)
+				}
+			}
+
+			By("Cleaning up cloud resources...")
+			framework.CleanupGCEIngressControllerWithTimeout(gceController, ingressesCleanupTimeout)
+		})
+
+		It("Create and update ingress should happen promptly with small/medium/large amount of ingresses", func() {
+			By(fmt.Sprintf("Bringing up %d backend servers...", scaleTestNumBackends))
+			testDeploy := generateScaleTestBackendDeploymentSpec(scaleTestNumBackends)
+			_, err := f.ClientSet.ExtensionsV1beta1().Deployments(ns).Create(testDeploy)
+			defer func() {
+				By(fmt.Sprintf("Deleting deployment %s...", testDeploy.Name))
+				if err := f.ClientSet.ExtensionsV1beta1().Deployments(ns).Delete(testDeploy.Name, nil); err != nil {
+					framework.Logf("Failed to delete deployment %s: %v", testDeploy.Name, err)
+				}
+			}()
+			framework.ExpectNoError(err)
+
+			By(fmt.Sprintf("Ensuring TLS secret %s...", scaleTestSecretName))
+			jig.PrepareTLSSecret(ns, scaleTestSecretName, scaleTestHostname)
+
+			currentNum := new(int)
+
+			prepareIngsFunc := func(goalNum int) {
+				var ingWg sync.WaitGroup
+				numToCreate := goalNum - *currentNum
+				ingWg.Add(numToCreate)
+				errQueue := make(chan error, numToCreate)
+				start := time.Now()
+				for ; *currentNum < goalNum; *currentNum++ {
+					suffix := fmt.Sprintf("%d", *currentNum)
+					go func() {
+						defer GinkgoRecover()
+						defer ingWg.Done()
+						acquireThread()
```

Comment:
Can you add a comment regarding what you are doing with acquire & release thread? Same with the sync.WaitGroup.